helix-term/src/application.rs

```diff
@@ -200,75 +213,107 @@ impl Application {
+        }
+    }
+
+    pub async fn event_loop_until_idle<S>(&mut self, input_stream: &mut S) -> bool
+    where
+        S: Stream<Item = crossterm::Result<crossterm::event::Event>> + Unpin,
+    {
+        loop {
+            if self.editor.should_close() {
+                return false;
+            }
+
             use futures_util::StreamExt;
 
             tokio::select! {
                 biased;
 
-                event = reader.next() => {
-                    self.handle_terminal_events(event)
+                Some(event) = input_stream.next() => {
+                    self.handle_terminal_events(event);
+                    self.editor.reset_idle_timer();
                 }
                 Some(signal) = self.signals.next() => {
                     self.handle_signals(signal).await;
+                    self.editor.reset_idle_timer();
                 }
                 Some((id, call)) = self.editor.language_servers.incoming.next() => {
                     self.handle_language_server_message(call, id).await;
                     // limit render calls for fast language server messages
                     let last = self.editor.language_servers.incoming.is_empty();
-                    if last || last_render.elapsed() > deadline {
+
+                    if last || self.last_render.elapsed() > LSP_DEADLINE {
                         self.render();
-                        last_render = Instant::now();
+                        self.last_render = Instant::now();
                     }
+
+                    self.editor.reset_idle_timer();
                 }
                 Some(payload) = self.editor.debugger_events.next() => {
                     let needs_render = self.editor.handle_debugger_message(payload).await;
                     if needs_render {
                         self.render();
                     }
+                    self.editor.reset_idle_timer();
                 }
                 Some(config_event) = self.editor.config_events.1.recv() => {
                     self.handle_config_events(config_event);
                     self.render();
+                    self.editor.reset_idle_timer();
                 }
                 Some(callback) = self.jobs.futures.next() => {
                     self.jobs.handle_callback(&mut self.editor, &mut self.compositor, callback);
                     self.render();
+                    self.editor.reset_idle_timer();
                 }
                 Some(callback) = self.jobs.wait_futures.next() => {
                     self.jobs.handle_callback(&mut self.editor, &mut self.compositor, callback);
                     self.render();
+                    self.editor.reset_idle_timer();
                 }
                 _ = &mut self.editor.idle_timer => {
                     // idle timeout
                     self.editor.clear_idle_timer();
                     self.handle_idle_timeout();
+                    return true;
                 }
             }
```

Comment:
Only had the time for a brief look, but I don't like the event loop changes. 

1) It adds an extra async function call to what was before a single loop (that's very frequently called)
2) Events & callbacks should not interact with the idle timer because it's intended to only be affected by keypresses. Continually resetting the idle timer on LSP events for example will ensure the timer never triggers for the first 10-20s while rust-analyzer boots up and continually spews log messages.

Alternate frontends will re-implement the application loop themselves, so maybe the solution is to treat the integration tests as a separate frontend? i.e. define a separate `test::Application` that has no rendering and declares it's own loop. I started working on this previously by moving the `handle_debugger_message` onto `helix-view`/`Editor`. We'd need to do the same for `handle_language_server_message` so we'd be able to share most of the code.

(Also, I move commands & parts of UI to helix-view on this branch: https://github.com/helix-editor/helix/tree/gui)

Reply:
> Only had the time for a brief look, but I don't like the event loop changes.
> 
>     1. It adds an extra async function call to what was before a single loop (that's very frequently called)

Could you please elaborate on what your concern is about this? Is it performance? Aren't we just talking about an extra jump that happens once every few hundred milliseconds in the worst case?

>     2. Events & callbacks should not interact with the idle timer because it's intended to only be affected by keypresses. Continually resetting the idle timer on LSP events for example will ensure the timer never triggers for the first 10-20s while rust-analyzer boots up and continually spews log messages.

I believe this is already the case. It's a biased select, so it will handle key presses, signals, LSP, etc, all before it polls the idle timer. I believe the only practical difference this makes is adding a few hundred ms once that initialization finishes.

Also, even if this were not the case, the LSP would not be able to serve completions until it's fully initialized anyway.

Perhaps we could make the reset calls conditional with the integration feature?

> Alternate frontends will re-implement the application loop themselves, so maybe the solution is to treat the integration tests as a separate frontend? i.e. define a separate `test::Application` that has no rendering and declares it's own loop.

This was my first thought when I ran into this problem as well, but then I considered that we want to be able to test all features, and I realized we would just be copy/pasting the entire loop with a few small tweaks. And this kind of defeats the point of integration testing, which is to test the very same code that the user will be running.

We want to be able to test things like `:wq` saves the file and exits the Application, in that order. Or doesn't exit when there's an error saving.

And we don't want to have to worry about updating the testing harness's event loop every time we touch the term Application's event loop.

Reply:
@archseer I consolidated the reset calls into a single place and put it behind the integration feature flag. This way, non test builds will see no difference in behavior; and additionally, since key presses are the only thing that triggers idle timer resets, this reduces the frequency that the call stack exits and reenters the inner event loop to once every idle timer fire, which should make the performance impact negligible. How to you feel about this?